File: erlang/zipper/src/zipper.erl
-module(zipper).

-export([from_tree/1, left/1, new_tree/3, right/1,
	 set_left/2, set_right/2, set_value/2, to_tree/1, up/1,
	 value/1]).

new_tree(V, L, R) -> {L, V, R}.

from_tree(Tree) -> {Tree, []}.

to_tree({Curr, []}) -> Curr;
to_tree(Zipper) -> to_tree(up(Zipper)).

value({{_, V, _}, _}) -> V.

set_value({{L, _, R}, Trail}, V) ->
    {{L, V, R}, Trail}.

set_left({{_, V, R}, Trail}, L) ->
    {{L, V, R}, Trail}.

set_right({{L, V, _}, Trail}, R) ->
    {{L, V, R}, Trail}.

left({{nil, _, _}, _}) -> nil;
% add Parent node to Tail instead just Value and RightChild for ease of understanding
left({Parent={L, _, _}, Trail}) ->
    {L, [{left, Parent} | Trail]}.

right({{_, _, nil}, _}) -> nil;
right({Parent={_, _, R}, Trail}) ->
    {R, [{right, Parent} | Trail]}.

up({_, []}) -> nil;
up({Curr, [{left, _Parent={_, V, R}} | T]}) ->
    {{Curr, V, R}, T};
up({Curr, [{right, _Parent={L, V, _}} | T]}) ->
    {{L, V, Curr}, T}.

File: haskell/zipper/src/Zipper.hs
module Zipper
  ( BinTree (BT),
    fromTree,
    left,
    right,
    setLeft,
    setRight,
    setValue,
    toTree,
    up,
    value,
  )
where

data BinTree a = BT
  { btValue :: a,
    btLeft :: Maybe (BinTree a),
    btRight :: Maybe (BinTree a)
  }
  deriving (Eq, Show)

data Direction = LF | RG deriving (Eq, Show)

data Zipper a = Z
  { zFocus :: (BinTree a),
    zRest :: [(a, Direction, Maybe (BinTree a))]
  }
  deriving (Eq, Show)

fromTree :: BinTree a -> Zipper a
fromTree tree = Z tree []

toTree :: Zipper a -> BinTree a
toTree z = zFocus $ top z

top :: Zipper a -> Zipper a
top z = case up z of
  Nothing -> z
  Just z' -> top z'

value :: Zipper a -> a
value (Z (BT v _ _) _) = v

left :: Zipper a -> Maybe (Zipper a)
left (Z (BT v l r) rest) = case l of
  Nothing -> Nothing
  Just l' -> Just $ Z l' ((v, LF, r) : rest)

right :: Zipper a -> Maybe (Zipper a)
right (Z (BT v l r) rest) = case r of
  Nothing -> Nothing
  Just r' -> Just $ Z r' ((v, RG, l) : rest)

up :: Zipper a -> Maybe (Zipper a)
up (Z _ []) = Nothing
up (Z focus ((v, direction, tree) : rest)) = Just $ Z (f direction) rest
  where
    f LF = BT v (Just focus) tree
    f RG = BT v tree (Just focus)

setValue :: a -> Zipper a -> Zipper a
setValue v (Z (BT _ l r) rest) = Z (BT v l r) rest

setLeft :: Maybe (BinTree a) -> Zipper a -> Zipper a
setLeft tree (Z (BT v _ r) rest) = (Z (BT v tree r) rest)

setRight :: Maybe (BinTree a) -> Zipper a -> Zipper a
setRight tree (Z (BT v l _) rest) = (Z (BT v l tree) rest)
